File: problems/0981/solution.py
class TimeMap:
    def __init__(self):
        self.__map: dict[str, list[tuple[int, str]]] = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.__map:
            self.__map[key] = []
        self.__map[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.__map:
            return ""
        tmp: list[tuple[int, str]] = self.__map.get(key)
        left = 0
        right = len(tmp) - 1
        while left <= right:
            mid = left + (right - left) // 2
            if tmp[mid][0] == timestamp:
                return tmp[mid][1]
            elif tmp[mid][0] < timestamp:
                left = mid + 1
            else:
                right = mid - 1
        return "" if right < 0 else tmp[right][1]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: problems/0981/solution.cpp
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class TimeMap {
  unordered_map<string, vector<pair<int, string>>> map;

 public:
  TimeMap() {}

  void set(string key, string value, int timestamp) {
    map[key].push_back({timestamp, value});
  }

  string get(string key, int timestamp) {
    auto it = map.find(key);
    if (it == map.end()) {
      return "";
    }
    auto& tmp = it->second;
    int left = 0;
    int right = tmp.size() - 1;
    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (tmp[mid].first == timestamp) {
        return tmp[mid].second;
      } else if (tmp[mid].first < timestamp) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    if (right < 0) {
      return "";
    }
    return tmp[right].second;
  }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */